src/control/lib/telemetry/telemetry.go

```diff
@@ -0,0 +1,396 @@
+	res := C.d_tm_get_metadata(&shortDesc, &longDesc, mb.handle.shmem, mb.node, C.CString(mb.Name()))
+	if res == C.DER_SUCCESS {
+		short := C.GoString(shortDesc)
+		mb.shortDesc = &short
+		long := C.GoString(longDesc)
+		mb.longDesc = &long
+
+		C.free(unsafe.Pointer(shortDesc))
+		C.free(unsafe.Pointer(longDesc))
+	} else {
+		failed := "failed to retrieve metadata"
+		mb.shortDesc = &failed
+		mb.longDesc = &failed
+	}
+}
+
+func (mb *metricBase) ShortDesc() string {
+	if mb.shortDesc == nil {
+		mb.fillMetadata()
+	}
+
+	return *mb.shortDesc
+}
+
+func (mb *metricBase) LongDesc() string {
+	if mb.longDesc == nil {
+		mb.fillMetadata()
+	}
+
+	return *mb.longDesc
+}
+
+func (mb *metricBase) String() string {
+	r, w, err := os.Pipe()
+	if err != nil {
+		return err.Error()
+	}
+	defer r.Close()
+	defer w.Close()
+
+	f := C.fdopen(C.int(w.Fd()), C.CString("w"))
+	if f == nil {
+		return "fdopen() failed"
+	}
+
+	go func() {
+		C.d_tm_print_node(mb.handle.shmem, mb.node, C.int(0), f)
+		C.fclose(f)
+	}()
+
+	buf := make([]byte, 128)
+
+	_, err = r.Read(buf)
+	if err != nil && err != io.EOF {
+		return err.Error()
+	}
+
+	return strings.TrimSpace(string(buf[:bytes.Index(buf, []byte{0})]))
+}
+
+func (sm *statsMetric) Mean() float64 {
+	return float64(sm.stats.mean)
+}
+
+func (sm *statsMetric) StdDev() float64 {
+	return float64(sm.stats.std_dev)
+}
+
+func (sm *statsMetric) SampleSize() uint64 {
+	return uint64(sm.stats.sample_size)
+}
+
+func getStatsMinInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_min_int(stats))
+}
+
+func getStatsMaxInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_max_int(stats))
+}
+
+func getStatsMinFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_min_float(stats))
+}
+
+func getStatsMaxFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_max_float(stats))
+}
+
+func getStatsSumInt(stats *C.struct_d_tm_stats_t) uint64 {
+	return uint64(C.get_sum_int(stats))
+}
+
+func getStatsSumFloat(stats *C.struct_d_tm_stats_t) float64 {
+	return float64(C.get_sum_float(stats))
+}
+
+func Init(parent context.Context, idx uint32) (context.Context, error) {
+	shmemRoot := C.d_tm_get_shared_memory(C.int(idx))
+	if shmemRoot == nil {
+		return nil, errors.Errorf("no shared memory segment found for idx: %d", idx)
+	}
+
+	root := C.d_tm_get_root(shmemRoot)
+	if root == nil {
+		return nil, errors.Errorf("no root node found in shared memory segment for idx: %d", idx)
+	}
+
+	handle := &handle{
+		idx:   idx,
+		shmem: shmemRoot,
+		root:  root,
+	}
+
+	return context.WithValue(parent, handleKey, handle), nil
+}
+
+func visit(hdl *handle, node *C.struct_d_tm_node_t, pathComps []string, out chan<- Metric) {
+	var next *C.struct_d_tm_node_t
+
+	if node == nil {
+		return
+	}
+	path := strings.Join(pathComps, "/")
+	name := C.GoString((*C.char)(C.d_tm_conv_ptr(hdl.shmem, unsafe.Pointer(node.dtn_name))))
+	//fmt.Printf("visiting: %q/%q (%p)\n", path, name, node)
+
```

Comment:
Similar to above comment, do you want to support the other metric types?

Reply:
My plan at this point is to implement what we need for Prometheus support and use that subset of functionality to refine the approach. Eventually we can add support for the other types to make it more of a complete solution, but I want to hold off on that for now because we don't need those types and I don't want to maintain/test them while iterating on the basic solution.